File: JavaScript/shellSort.js
/**
 * Function to perform Shell Sort on an array.
 * @param {number[]} arr - The array to be sorted.
 * @returns {number[]} - The sorted array.
 */
function shellSort(arr) {
    let n = arr.length; // Get the length of the array

    // Start with a gap equal to half the length of the array
    for (let gap = Math.floor(n / 2); gap > 0; gap = Math.floor(gap / 2)) {
        // Perform a gapped insertion sort for this gap size
        for (let i = gap; i < n; i++) {
            let temp = arr[i]; // Store the current element
            let j; // Initialize the index for comparison

            // Shift earlier gap-sorted elements up until the correct location is found
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                arr[j] = arr[j - gap]; // Move the larger element to the current index
            }

            // Put the stored element into its correct location
            arr[j] = temp;
        }
    }

    return arr; // Return the sorted array
}

// Example usage
const arrayToSort = [64, 34, 25, 12, 22, 11, 90];
const sortedArray = shellSort(arrayToSort);
console.log("Sorted Array:", sortedArray);

File: Python/shellSort.py
def shell_sort(arr):
    
    n = len(arr)  # Get the length of the array
    gap = n // 2  # Initialize gap size, starting with half the array length

    # Keep reducing the gap until it becomes zero
    while gap > 0:
        # Perform a gapped insertion sort for this gap size
        for i in range(gap, n):
            temp = arr[i]  # Store the current element
            j = i  # Start with the current position
            
            # Shift earlier gap-sorted elements up until the correct location for temp is found
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]  # Move the larger element to the current position
                j -= gap  # Move to the previous position with the same gap
            
            arr[j] = temp  # Insert the element into its correct position
            
        gap //= 2  # Reduce the gap size

    return arr  # Return the sorted array

# Example usage
if __name__ == "__main__":
    # Sample array to sort
    sample_array = [12, 34, 54, 2, 3]
    print("Original array:", sample_array)
    
    # Sorting the array
    sorted_array = shell_sort(sample_array)
    print("Sorted array:", sorted_array)
